File: [0069][Easy][Sqrt(x)]/Sqrt(x)_3.py
# solution 3, bit op and recursion,
# leetcode time     cost : 40 ms
# leetcode memory   cost : 13.1 MB 
# Time  Complexity: O(logN)
# Space Complexity: O(logN)
class Solution:
    def mySqrt(self, x):
        if x < 2:
            return x
        
        left = self.mySqrt(x >> 2) << 1
        right = left + 1
        return left if right * right > x else right

File: [0069][Easy][Sqrt(x)]/Sqrt(x)_3.java
// solution 3, bit op and recursion,
// leetcode time     cost : 1 ms
// leetcode memory   cost : 39.9 MB 
// Time  Complexity: O(logN)
// Space Complexity: O(logN)
class Solution {
  public int mySqrt(int x) {
    if (x < 2) return x;

    int left = mySqrt(x >> 2) << 1;
    int right = left + 1;
    return (long)right * right > x ? left : right;
  }
}